be/dtm0_log.h

```diff
@@ -0,0 +1,232 @@
+ *    log record: a log record with empty payload (NULL). The users of DTM0 log
+ *    should be ready to encounter such a log entry, and treat it differently
+ *    whenever it is required.
+ *
+ * {
+ *    struct m0_be_tx_credit     cred;
+ *    struct m0_be_dtm0_log     *log;
+ *    struct m0_be_tx           *tx;
+ *    struct m0_dtm0_log_record *txr;
+ *    struct m0_be_seg          *seg;
+ *
+ *    m0_be_dtm0_log_credit(op, tx, seg, &cred);
+ *    m0_be_tx_prep(tx, &cred);
+ *    m0_be_tx_open(tx);
+ *
+ *    ...
+ *    m0_be_dtm0_log_update(log, tx, txr);
+ *    tx_close(tx);
+ * }
+ *
+ *
+ */
+
+/* Participant state */
+enum m0_dtm0_log_pa_state {
+	M0_DTML_STATE_INPROGRESS,
+	M0_DTML_STATE_EXECUTED,
+	M0_DTML_STATE_PERSISTENT,
+};
+
+enum m0_be_dtm0_log_credit_op {
+	M0_DTML_CREATE,
+	M0_DTML_SENT,
+	M0_DTML_EXECUTED,
+	M0_DTML_PERSISTENT,
+	M0_DTML_REDO
+};
+
+/* Unique identifier for request */
+struct m0_dtm0_dtx_id {
+	struct m0_fid fid;
+	uint64_t      ts;
+};
+
+struct m0_dtm0_log_pa {
+	struct m0_fid             pfid;
+	enum m0_dtm0_log_pa_state pstate;
+};
+
+/* TODO: define in a separate header. */
+struct m0_dtm0_txr {
+	struct m0_dtm0_dtx_id  dt_tid;
+	struct m0_dtm0_log_pa *dt_participants;
+	uint16_t               dt_participants_nr;
+	struct m0_buf          dt_txr_payload;
+};
+
+struct m0_dtm0_log_record {
+	uint64_t               dlr_magic;
+	struct m0_dtm0_txr     dlr_txr;
+	struct m0_be_list_link dlr_link; /* link into m0_be_dtm0_log::list */
+	struct m0_list_link    dlr_tlink;
+};
+
+struct m0_be_dtm0_log {
+	struct m0_mutex    dl_lock;  /* volatile structure */
+	struct m0_be_list *dl_list;  /* persistent structure */
+	struct m0_list    *dl_vlist; /* Volatile list */
+};
+
+// init/fini (for volatile fields)
+M0_INTERNAL void m0_be_dtm0_log_init(struct m0_be_dtm0_log *log);
+M0_INTERNAL void m0_be_dtm0_log_fini(struct m0_be_dtm0_log *log);
+
+// credit interface
+M0_INTERNAL void m0_be_dtm0_log_credit(enum m0_be_dtm0_log_credit_op op,
+                                       struct m0_be_tx              *tx,
+                                       struct m0_be_seg             *seg,
+                                       struct m0_be_tx_credit       *accum);
+// create/destroy
+M0_INTERNAL struct m0_be_dtm0_log *m0_be_dtm0_log_create(struct m0_be_tx i *tx,
+                                                         struct m0_be_seg *seg);
+
+M0_INTERNAL void m0_be_dtm0_log_destroy(struct m0_be_dtm0_log **log,
+                                        struct m0_be_tx        *tx);
+
+// operational interfaces
+M0_INTERNAL void m0_be_dtm0_log_update(struct m0_be_dtm0_log      *log,
+                                       struct m0_be_tx            *tx,
+                                       struct m0_dtm0_txr         *txr);
+
+M0_INTERNAL void m0_be_dtm0_log_find(struct m0_be_dtm0_log *log,
+                                     struct m0_dtm0_dtx_id *id,
```

Comment:
Suggestion: let's make it constant (`const struct m0_dtm0_dtx_id *id`).

Rationale: this structure does not have (and should never have) any dynamic parts (for example, mutexes), so that we can freely add "const" here.

Reply:
Addressed at commit a3b0fecd62dfa787c7177d1487fefcb3e45ed47a